indexer/tasks/src/multiera/multiera_projected_nft.rs

```diff
@@ -0,0 +1,373 @@
+    db_tx: &DatabaseTransaction,
+    multiera_used_inputs_to_outputs_map: &BTreeMap<Vec<u8>, BTreeMap<i64, OutputWithTxData>>,
+) -> Result<BTreeMap<Vec<u8>, BTreeMap<i64, ProjectedNftInputsQueryOutputResult>>, DbErr> {
+    let inputs_condition = multiera_used_inputs_to_outputs_map
+        .iter()
+        .flat_map(|(_input_tx_id, map)| {
+            map.iter().map(|(_output_index, data)| {
+                (data.model.tx_id, data.model.output_index, data.model.id) // input and utxo id
+            })
+        })
+        .fold(Condition::any(), |cond, (tx_id, _output_index, utxo_id)| {
+            cond.add(
+                Condition::all()
+                    .add(ProjectedNftColumn::HololockerUtxoId.eq(utxo_id))
+                    .add(ProjectedNftColumn::TxId.eq(tx_id)),
+            )
+        });
+
+    let projected_nfts = ProjectedNft::find()
+        .select_only()
+        .column(TransactionOutputColumn::Id)
+        .column(TransactionOutputColumn::TxId)
+        .column(TransactionOutputColumn::OutputIndex)
+        .column(ProjectedNftColumn::Operation)
+        .column(ProjectedNftColumn::Asset)
+        .column(ProjectedNftColumn::Amount)
+        .column(ProjectedNftColumn::OwnerAddress)
+        .column_as(TransactionColumn::Hash, "tx_hash")
+        .join(
+            JoinType::InnerJoin,
+            ProjectedNftRelation::TransactionOutput.def(),
+        )
+        .join(JoinType::InnerJoin, ProjectedNftRelation::Transaction.def())
+        .filter(inputs_condition)
+        .into_model::<ProjectedNftInputsQueryOutputResult>()
+        .all(db_tx)
+        .await?;
+
+    let mut result: BTreeMap<Vec<u8>, BTreeMap<i64, ProjectedNftInputsQueryOutputResult>> =
+        BTreeMap::new();
+    for nft in projected_nfts {
+        result
+            .entry(nft.tx_hash.clone())
+            .or_default()
+            .entry(nft.output_index as i64)
+            .or_insert(nft);
+    }
+    Ok(result)
+}
+
+#[derive(Debug, Clone)]
+struct ProjectedNftData {
+    pub previous_utxo_tx_hash: Vec<u8>,
+    pub previous_utxo_tx_output_index: Option<i64>,
+    pub address: Vec<u8>,
+    pub plutus_data: Vec<u8>,
+    pub operation: ProjectedNftOperation,
+    pub for_how_long: Option<i64>,
+}
+
+fn extract_operation_and_datum(output: &MultiEraOutput) -> ProjectedNftData {
+    let datum_option = match output.datum() {
+        Some(datum) => DatumOption::from(datum.clone()),
+        None => {
+            return ProjectedNftData {
+                previous_utxo_tx_hash: vec![],
+                previous_utxo_tx_output_index: None,
+                address: vec![],
+                plutus_data: vec![],
+                operation: ProjectedNftOperation::NoDatum,
+                for_how_long: None,
+            };
+        }
+    };
+
+    let datum = match datum_option {
+        DatumOption::Hash(hash) => {
+            return ProjectedNftData {
+                previous_utxo_tx_hash: vec![],
```

Comment:
We can probably use `..Default::default()` to shrink the code for a lot of these